src/rust/src/x509/sign.rs

```diff
@@ -256,3 +258,142 @@ pub(crate) fn sign_data<'p>(
+) -> PyAsn1Result<&'p pyo3::PyAny> {
+    let key_type = identify_public_key_type(py, issuer_public_key)?;
+    let sig_key_type = identify_key_type_for_oid(signature_oid)?;
+    if key_type != sig_key_type {
+        return Err(PyAsn1Error::from(pyo3::exceptions::PyValueError::new_err(
+            "Signature algorithm does not match issuer key type",
+        )));
+    }
+
+    match key_type {
+        KeyType::Ed25519 | KeyType::Ed448 => {
+            issuer_public_key.call_method1("verify", (signature, data))?
+        }
+        KeyType::Ec => {
+            let ec_mod = py.import("cryptography.hazmat.primitives.asymmetric.ec")?;
+            let ecdsa = ec_mod
+                .getattr(crate::intern!(py, "ECDSA"))?
+                .call1((signature_hash,))?;
+            issuer_public_key.call_method1("verify", (signature, data, ecdsa))?
+        }
+        KeyType::Rsa => {
+            let padding_mod = py.import("cryptography.hazmat.primitives.asymmetric.padding")?;
+            let pkcs1v15 = padding_mod
+                .getattr(crate::intern!(py, "PKCS1v15"))?
+                .call0()?;
+            issuer_public_key.call_method1("verify", (signature, data, pkcs1v15, signature_hash))?
+        }
+        KeyType::Dsa => {
+            issuer_public_key.call_method1("verify", (signature, data, signature_hash))?
+        }
+    };
+    Ok(py.None().into_ref(py))
+}
+
+fn identify_public_key_type(
+    py: pyo3::Python<'_>,
+    public_key: &pyo3::PyAny,
+) -> pyo3::PyResult<KeyType> {
+    let rsa_key_type: &pyo3::types::PyType = py
+        .import("cryptography.hazmat.primitives.asymmetric.rsa")?
+        .getattr(crate::intern!(py, "RSAPublicKey"))?
+        .extract()?;
+    let dsa_key_type: &pyo3::types::PyType = py
+        .import("cryptography.hazmat.primitives.asymmetric.dsa")?
+        .getattr(crate::intern!(py, "DSAPublicKey"))?
+        .extract()?;
+    let ec_key_type: &pyo3::types::PyType = py
+        .import("cryptography.hazmat.primitives.asymmetric.ec")?
+        .getattr(crate::intern!(py, "EllipticCurvePublicKey"))?
+        .extract()?;
+    let ed25519_key_type: &pyo3::types::PyType = py
+        .import("cryptography.hazmat.primitives.asymmetric.ed25519")?
+        .getattr(crate::intern!(py, "Ed25519PublicKey"))?
+        .extract()?;
+    let ed448_key_type: &pyo3::types::PyType = py
+        .import("cryptography.hazmat.primitives.asymmetric.ed448")?
+        .getattr(crate::intern!(py, "Ed448PublicKey"))?
+        .extract()?;
+
+    if rsa_key_type.is_instance(public_key)? {
+        Ok(KeyType::Rsa)
+    } else if dsa_key_type.is_instance(public_key)? {
+        Ok(KeyType::Dsa)
+    } else if ec_key_type.is_instance(public_key)? {
+        Ok(KeyType::Ec)
+    } else if ed25519_key_type.is_instance(public_key)? {
+        Ok(KeyType::Ed25519)
+    } else if ed448_key_type.is_instance(public_key)? {
+        Ok(KeyType::Ed448)
+    } else {
+        Err(pyo3::exceptions::PyTypeError::new_err(
```

Comment:
Coverage failing because this is unreachable from Python code (`signature_hash_algorithm` calls catch any error before this can be invoked). I'm structuring this incorrectly clearly 😄